src/Sparrow/Utils/ZstdStream.cs

```diff
@@ -0,0 +1,361 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Sparrow.Threading;
+
+namespace Sparrow.Utils
+{
+    internal class ZstdStream : Stream
+    {
+        private Stream _inner;
+        private readonly bool _compression;
+        private ZstdLib.CompressContext _compressContext = new ZstdLib.CompressContext();
+        private byte[] _tempBuffer = ArrayPool<byte>.Shared.Rent(1024);
+        private Memory<byte> _decompressionInput = Memory<byte>.Empty;
+        private readonly DisposeOnceAsync<SingleAttempt> _disposeOnce;
+        private readonly SemaphoreSlim _readSemaphore = new SemaphoreSlim(1);
+        private readonly SemaphoreSlim _writeSemaphore = new SemaphoreSlim(1);
+        private ZstdStream(Stream inner, bool compression)
+        {
+            _inner = inner;
+            _compression = compression;
+            _disposeOnce = new DisposeOnceAsync<SingleAttempt>(DisposeInternal);
+        }
+
+        public static ZstdStream Compress(Stream stream) => new ZstdStream(stream, compression: true);
+        public static ZstdStream Decompress(Stream stream) => new ZstdStream(stream, compression: false);
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
```

Comment:
`CanRead` and `CanWrite` should take their cue from `_compression`

Reply:
Done